Add new cli for bf lg:analyze

feich-ms opened the pull request:
Fix #1109. Support lg analyze cli functionality

chrimc62 commented on packages/cli/README.md:
```diff
@@ -460,6 +461,24 @@ OPTIONS
+  -h, --help     lg:analyze help
+  -i, --in=in    (required) LG File or folder that contains .lg file(s)
+  -o, --out=out  Output file or folder name. If not specified stdout will be used as output
+  -r, --recurse  Considere sub-folders to find .lg file(s)
```
>Considere [](start = 17, length = 9)

Consider

feich-ms replied:
resolved

chrimc62 commented on packages/lg/README.md:
```diff
@@ -96,6 +97,24 @@ OPTIONS
+  -h, --help     lg:analyze help
+  -i, --in=in    (required) LG File or folder that contains .lg file(s)
+  -o, --out=out  Output file or folder name. If not specified stdout will be used as output
+  -r, --recurse  Considere sub-folders to find .lg file(s)
```
>Considere [](start = 17, length = 9)

Spelling

feich-ms replied:
resolved

chrimc62 commented:
>   -r, --recurse  Considere sub-folders to find .lg file(s)

Spelling

---
Refers to: packages/lg/README.md:96 in 4da2346. [](commit_id = 4da23467daa18c24f7fdc04c4f042d1b58296d61, deletion_comment = False)

chrimc62 reviewed (approved):
:shipit:

chrimc62 commented:
Thanks for doing this!  I have definitely found this useful for more complex template patterns.

joshgummersall commented on packages/lg/src/commands/lg/analyze.ts:
```diff
@@ -0,0 +1,165 @@
+type Source = string
+type SourceToReferences = Map<Source, TemplateName[]>
+type TemplateToReferences = Map<FullTemplateName, SourceToReferences>
+const SourceToReferences = <{ new(): SourceToReferences }>Map
```
I've not seen this syntax before - what is the intent? Instead of this, change `new SourceToReferences()` to `const value: SourceToReferences = new Map()`.

feich-ms replied:
Hi @joshgummersall, thanks for you suggestions. I just removed these syntax from code. It seems they are not needed at all. These syntax comes from Chris McConnell https://github.com/microsoft/botframework-cli/issues/1109#issue-802627455, it seems defining the constructor of type SourceToReferences. With this kind of definition, we can use `new TemplateToReferences() ` to new a map of type TemplateToReferences. Without this, we can instead use `new Map<Source, TemplateName[]>())`. I prefer the latter which is more readable and friendly to developers.

feich-ms commented:
@munozemilio thanks for your review and suggestions about above syntax issues. I have refactored that and removed such syntax that is not commonly used. Could you help to approve and merge? Thanks.

munozemilio commented on packages/lg/src/commands/lg/analyze.ts:
```diff
@@ -0,0 +1,163 @@
+    const errors = diagnostics.filter(u => u.severity === DiagnosticSeverity.Error)
+    if (errors && errors.length > 0) {
+      throw new CLIError(errors.map(u => u.toString()).join('\n'))
+    } else {
```
Else not needed

feich-ms replied:
yes, good catch, removed else {}
